src/libs/elektra/key.c

```diff
@@ -341,125 +341,85 @@ Key * keyCopy (Key * dest, const Key * source, elektraCopyFlags flags)
-			dest->ukey = elektraMemDup (source->ukey, source->keyUSize);
-			if (!dest->ukey) goto memerror;
-			dest->keyUSize = source->keyUSize;
+			dest->keyName = source->keyName;
+			keyNameRefInc (dest->keyName);
 		}
 		else
 		{
-			dest->key = elektraStrDup ("/");
-			dest->keySize = 2;
-
-			dest->ukey = elektraMalloc (3);
-			dest->ukey[0] = KEY_NS_CASCADING;
-			dest->ukey[1] = '\0';
-			dest->ukey[2] = '\0';
-			dest->keyUSize = 3;
+			keySetName (dest, "/");
 		}
+
 		clear_bit (dest->flags, KEY_FLAG_MMAP_KEY);
 	}
 
-	if (test_bit (flags, KEY_CP_STRING))
+	if (test_bit (flags, KEY_CP_STRING) || test_bit (flags, KEY_CP_VALUE))
 	{
-		if (source->data.v != NULL)
+		if (dest->keyData != NULL)
 		{
-			dest->data.v = elektraMemDup (source->data.v, source->dataSize);
-			if (!dest->data.v) goto memerror;
-			dest->dataSize = source->dataSize;
-
-			if (!test_bit (flags, KEY_CP_META) && keyIsBinary (source))
-			{
-				keySetMeta (dest, "binary", "");
-			}
+			keyDataRefDecAndDel (dest->keyData, !test_bit (dest->flags, KEY_FLAG_MMAP_DATA));
+			dest->keyData = NULL;
 		}
-		else
-		{
-			dest->data.v = NULL;
-			dest->dataSize = 0;
-		}
-		clear_bit (dest->flags, KEY_FLAG_MMAP_DATA);
-	}
 
-	if (test_bit (flags, KEY_CP_VALUE))
-	{
-		if (source->data.v != NULL)
+		if (source->keyData != NULL)
 		{
-			dest->data.v = elektraMemDup (source->data.v, source->dataSize);
-			if (!dest->data.v) goto memerror;
-			dest->dataSize = source->dataSize;
+			dest->keyData = source->keyData;
+			keyDataRefInc (dest->keyData);
 
 			if (!test_bit (flags, KEY_CP_META) && keyIsBinary (source))
 			{
 				keySetMeta (dest, "binary", "");
 			}
 		}
-		else
-		{
-			dest->data.v = NULL;
-			dest->dataSize = 0;
-		}
+
 		clear_bit (dest->flags, KEY_FLAG_MMAP_DATA);
 	}
 
 	if (test_bit (flags, KEY_CP_META))
 	{
+		ksDel (dest->meta);
+
 		if (source->meta != NULL)
 		{
 			dest->meta = ksDup (source->meta);
 			if (!dest->meta) goto memerror;
 		}
 		else
 		{
-			dest->meta = 0;
+			dest->meta = NULL;
 		}
 	}
 
 	// successful, now do the irreversible stuff: we obviously modified dest
 	set_bit (dest->flags, KEY_FLAG_SYNC);
 
-	// free old resources of destination
-	if (test_bit (flags, KEY_CP_NAME) && !test_bit (orig.flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.key);
-	if (test_bit (flags, KEY_CP_NAME) && !test_bit (orig.flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.ukey);
-	if (test_bit (flags, KEY_CP_VALUE) && !test_bit (orig.flags, KEY_FLAG_MMAP_DATA))
-	{
-		elektraFree (orig.data.v);
-	}
-	else if (test_bit (flags, KEY_CP_STRING) && !test_bit (orig.flags, KEY_FLAG_MMAP_DATA))
-	{
-		elektraFree (orig.data.c);
-	}
-
-	if (test_bit (flags, KEY_CP_META)) ksDel (orig.meta);
-
 	return dest;
 
 memerror:
-	elektraFree (dest->key);
-	elektraFree (dest->data.v);
+	keyNameDel (dest->keyName, !test_bit (dest->flags, KEY_FLAG_MMAP_KEY));
+	keyDataDel (dest->keyData, !test_bit (dest->flags, KEY_FLAG_MMAP_DATA));
 	ksDel (dest->meta);
 
-	*dest = orig;
```

Comment:
This line was crucial... Together with having all the `elektraFree` _after_ the last `goto memerror`, it meant that `dest` would not be modified, if `keyCopy` fails.

The current version will corrupt `dest` on failure.

Reply:
You're right. I've added the functionality back in and also created a test that checks exactly that.